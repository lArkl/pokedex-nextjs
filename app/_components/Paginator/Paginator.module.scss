.container {
  display: flex;
  gap: 1rem;
  align-items: center;
  margin: 1rem 0;
  width: 100%;

  .page {
    width: 3rem;
    background-color: var(--color-purple-6);
    &:hover {
      background-color: var(--color-purple-5);
    }
  }

  .dots {
    font-size: large;
  }

  .arrow {
    background-color: var(--color-gray-2);
    color: white;
    &:hover {
      background-color: var(--color-gray-3);
    }

    &:disabled {
      background-color: var(--color-gray-4);
    }
  }

  .current {
    background-color: var(--color-purple-2);
  }
}
