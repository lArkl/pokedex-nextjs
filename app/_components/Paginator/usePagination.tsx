import { useMemo } from 'react'

interface UsePaginationProps {
  currentPage: number
  totalPages: number
  siblingCount?: number
}
const getIndexRange = (start: number, end: number): number[] =>
  Array.from({ length: end - start + 1 }, (_, idx) => idx + start)

const DOTS = -1000

export const usePagination = ({ currentPage, totalPages, siblingCount = 1 }: UsePaginationProps) => {
  const paginationRange = useMemo(() => {
    const pageNumbers = 5 + siblingCount

    if (pageNumbers >= totalPages) {
      return getIndexRange(1, totalPages)
    }

    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1)
    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPages)

    const showLeftDots = leftSiblingIndex > 2
    const showRightDots = rightSiblingIndex + 2 < totalPages

    const firstPageIndex = 1

    if (!showLeftDots && showRightDots) {
      const leftCount = 3 + 2 * siblingCount
      return [...getIndexRange(1, leftCount), DOTS, totalPages]
    }

    if (showLeftDots && !showRightDots) {
      const rightCount = 3 + 2 * siblingCount
      return [firstPageIndex, DOTS, ...getIndexRange(totalPages - rightCount + 1, totalPages)]
    }

    return [firstPageIndex, DOTS, ...getIndexRange(leftSiblingIndex, rightSiblingIndex), DOTS, totalPages]
  }, [siblingCount, totalPages, currentPage])

  return paginationRange
}

export const needsDots = (index: number | string): boolean => index === DOTS
