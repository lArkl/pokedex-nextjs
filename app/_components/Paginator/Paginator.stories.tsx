import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Paginator, { PaginatorProps } from './Paginator'
import { useState } from 'react'

const meta = {
  title: 'Molecules/Paginator',
  component: Paginator,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    setCurrentPage: {
      control: false,
    },
    currentPage: {
      control: false,
    },
  },
} satisfies Meta<typeof Paginator>

export default meta

const Component = ({ pageSize, totalCount }: Pick<PaginatorProps, 'pageSize' | 'totalCount'>) => {
  const [currentPage, setCurrentPage] = useState(1)
  return (
    <Paginator
      currentPage={currentPage}
      pageSize={pageSize}
      totalCount={totalCount}
      setCurrentPage={(page) => {
        action('setCurrentPage')(page)
        setCurrentPage(page)
      }}
    />
  )
}

type BaseStory = StoryObj<typeof meta>
type Story = Omit<BaseStory, 'args'> & { args: Omit<BaseStory['args'], 'setCurrentPage' | 'currentPage'> }

export const Default: Story = {
  args: { pageSize: 1, totalCount: 10 },
  render: (args) => <Component pageSize={args.pageSize} totalCount={args.totalCount} />,
}
