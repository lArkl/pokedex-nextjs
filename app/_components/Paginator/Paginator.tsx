"use client";

import { FC } from "react";
import styles from "./Paginator.module.scss";
import classNames from "classnames";
import { usePagination, needsDots } from "./usePagination";
import Button from "../Button";
import Typography from "../Typography";
import { usePathname, useSearchParams, useRouter } from "next/navigation";

export interface PaginatorProps {
  pageSize: number;
  currentPage: number;
  totalCount: number;
  className?: string;
}

const Paginator: FC<PaginatorProps> = ({
  currentPage,
  pageSize,
  totalCount,
  className,
}: PaginatorProps) => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();

  const setCurrentPage = (page: number): void => {
    const params = new URLSearchParams(searchParams);
    params.set("page", page.toString());
    replace(`${pathname}?${params.toString()}`);
  };

  const totalPages = Math.ceil(totalCount / pageSize);
  const indexRange = usePagination({ currentPage, totalPages });
  return (
    <ul className={classNames(styles.container, className)}>
      <li className={styles.item}>
        <Button
          className={styles.arrow}
          disabled={currentPage <= 1}
          onClick={() => {
            setCurrentPage(currentPage - 1);
          }}
        >
          {"<"}
        </Button>
      </li>
      {indexRange.map((pageIndex) => {
        const isDots = needsDots(pageIndex);
        return (
          <li key={`page-${pageIndex}`} className={styles.item}>
            {isDots ? (
              <Typography className={styles.dots}>...</Typography>
            ) : (
              <Button
                className={classNames(styles.page, {
                  [styles.current]: currentPage === pageIndex,
                })}
                onClick={() => setCurrentPage(pageIndex)}
              >
                {pageIndex}
              </Button>
            )}
          </li>
        );
      })}
      <li className={styles.item}>
        <Button
          className={styles.arrow}
          disabled={currentPage >= totalPages}
          onClick={() => {
            setCurrentPage(currentPage + 1);
          }}
        >
          {">"}
        </Button>
      </li>
    </ul>
  );
};

export default Paginator;
