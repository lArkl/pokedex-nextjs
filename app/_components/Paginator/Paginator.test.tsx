import { fireEvent, render, screen, within } from '@testing-library/react'
import Paginator, { PaginatorProps } from './Paginator'
import { vi } from 'vitest'

const setCurrentPageMock = vi.fn()

const renderComponent = (props?: Partial<PaginatorProps>) =>
  render(<Paginator currentPage={1} totalCount={15} pageSize={5} setCurrentPage={setCurrentPageMock} {...props} />)

describe('Paginator', () => {
  it('renders component', async () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })
  it('renders page items', async () => {
    renderComponent()

    expect(screen.getByRole('list')).toBeInTheDocument()
    expect(screen.getAllByRole('listitem')).toHaveLength(5)
  })

  it('calls setCurrentPage', async () => {
    renderComponent()

    const pageItem = screen.getAllByRole('listitem')[2]
    const pageButton = within(pageItem).getByRole('button', { name: '2' })

    fireEvent.click(pageButton)

    expect(setCurrentPageMock).toHaveBeenCalledWith(2)
  })
})
