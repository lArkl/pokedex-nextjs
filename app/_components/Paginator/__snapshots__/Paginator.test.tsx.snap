// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Paginator > renders component 1`] = `
<DocumentFragment>
  <ul
    class="_container_f9550d"
  >
    <li
      class="_item_f9550d"
    >
      <button
        class="_container_f8f296 _primary_f8f296 _medium_f8f296 _arrow_f9550d"
        disabled=""
      >
        &lt;
      </button>
    </li>
    <li
      class="_item_f9550d"
    >
      <button
        class="_container_f8f296 _primary_f8f296 _medium_f8f296 _page_f9550d _current_f9550d"
      >
        1
      </button>
    </li>
    <li
      class="_item_f9550d"
    >
      <button
        class="_container_f8f296 _primary_f8f296 _medium_f8f296 _page_f9550d"
      >
        2
      </button>
    </li>
    <li
      class="_item_f9550d"
    >
      <button
        class="_container_f8f296 _primary_f8f296 _medium_f8f296 _page_f9550d"
      >
        3
      </button>
    </li>
    <li
      class="_item_f9550d"
    >
      <button
        class="_container_f8f296 _primary_f8f296 _medium_f8f296 _arrow_f9550d"
      >
        &gt;
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
